dace/codegen/targets/cpu.py

```diff
@@ -1098,15 +1209,23 @@ def process_out_memlets(self,
                             write_expr = f"*({ptr_str} + {array_expr}) = {in_local_name};"
                         else:
                             desc_dtype = desc.dtype
-                            expr = cpp.cpp_array_expr(sdfg, memlet, codegen=self._frame)
+                            # If the storage type if CPU_Heap or GPU_Global then it might be requiring deferred allocation
+                            # We can check if the array requires sepcial access using A_size[0] (CPU) or __A_dim0_size (GPU0)
+                            # by going through the shape and checking for symbols starting with __dace_defer
+                            deferred_size_names = self._get_deferred_size_names(desc, memlet)
+                            expr = cpp.cpp_array_expr(sdfg, memlet, codegen=self._frame, deferred_size_names=deferred_size_names)
```

Comment:
To make the API cleaner, would it make sense to move `_get_deferred_size_names` inside `cpp` or `cpp.cpp_array_expr`?

Reply:
I moved this function call to cpp as much as I can and also try to hide it as much as I can from function signatures.